File: go/LUIS/quickstart.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"github.com/Azure/azure-sdk-for-go/services/cognitiveservices/v2.0/luis/authoring"
	"github.com/Azure/go-autorest/autorest"
	"github.com/satori/go.uuid"
	"log"
	"os"
	"strings"
	"time"
)

/* This sample for the Azure Cognitive Services QnA Maker API shows how to:
 * - Create an application.
 * - Add intents to an application.
 * - Add entities to an application.
 * - Add utterances to an application.
 * - Train an application.
 * - Publish an application.
 * - Delete an application.
 * - List all applications.
 */
 
/*  Configure the local environment:
	* Set the LUIS_SUBSCRIPTION_KEY and LUIS_REGION environment variables 
	* on your local machine using the appropriate method for your preferred shell 
	* (Bash, PowerShell, Command Prompt, etc.). 
	*
	* For LUIS_REGION, use the same region you used to get your subscription keys. 
	*
	* If the environment variable is created after the application is launched in a console or with Visual
	* Studio, the shell (or Visual Studio) needs to be closed and reloaded to take the environment variable into account.
	*/
var subscription_key string = os.Getenv("LUIS_SUBSCRIPTION_KEY")
var region string = os.Getenv("LUIS_REGION")

// Replace this with the endpoint for your subscription key.
var endpoint string = "https://" + region + ".api.cognitive.microsoft.com"
//  END - Configure the local environment.

func create_app() (string, string) {
	// Create a new LUIS app
	// Get the context, which is required by the SDK methods.
	ctx := context.Background()

	client := authoring.NewAppsClient(endpoint)
	// Set the subscription key on the client.
	client.Authorizer = autorest.NewCognitiveServicesAuthorizer(subscription_key)

	name := "Contoso " + time.Now().Format("2006-01-02 15:04:05")
	description := "Flight booking app built with Azure SDK for Go."
	version := "0.1"
	locale  := "en-us"

	create_app_payload := authoring.ApplicationCreateObject { Name: &name, Description: &description, InitialVersionID: &version, Culture: &locale }

	result, err := client.Add(ctx, create_app_payload)
	if err != nil {
		log.Fatal(err)
	}

	fmt.Printf("Created LUIS app %s with ID %s\n", name, (*result.Value).String())
	return (*result.Value).String(), version
}

func add_entities (app_id string, app_version string) {
// Get the context, which is required by the SDK methods.
	ctx := context.Background()

	client := authoring.NewModelClient(endpoint)
	// Set the subscription key on the client.
	client.Authorizer = autorest.NewCognitiveServicesAuthorizer(subscription_key)

	uuid_app_id, err := uuid.FromString(app_id)
    if err != nil {
		log.Fatal(err)
    }
	entity_name := "Destination"
	_, err = client.AddEntity(ctx, uuid_app_id, app_version, authoring.ModelCreateObject { Name: &entity_name })
    if err != nil {
		log.Fatal(err)
    }
	
	h_entity_name := "Class"
	h_entity_children := []string{ "First", "Business", "Economy" }
	_, err = client.AddHierarchicalEntity(ctx, uuid_app_id, app_version, authoring.HierarchicalEntityModel { Name: &h_entity_name, Children: &h_entity_children })
    if err != nil {
		log.Fatal(err)
    }

	c_entity_name := "Flight"
	c_entity_children := []string{ "Class", "Destination" }
	_, err = client.AddCompositeEntity(ctx, uuid_app_id, app_version, authoring.CompositeEntityModel { Name: &c_entity_name, Children: &c_entity_children })
	    if err != nil {
		log.Fatal(err)
    }

	fmt.Println("Entities Destination, Class, Flight created.")
}

func add_intents(app_id string, app_version string) {
	ctx := context.Background()

	client := authoring.NewModelClient(endpoint)
	// Set the subscription key on the client.
	client.Authorizer = autorest.NewCognitiveServicesAuthorizer(subscription_key)

	uuid_app_id, err := uuid.FromString(app_id)
    if err != nil {
		log.Fatal(err)
    }

	name := "FindFlights"
	_, err = client.AddIntent(ctx, uuid_app_id, app_version, authoring.ModelCreateObject { Name: &name })

	fmt.Println("Intent FindFlights added.")
}

func create_utterance(intent string, text string, labels map[string]string) authoring.ExampleLabelObject {
    text = strings.ToLower(text)

	var results []authoring.EntityLabelObject

	for key, value := range labels {
		/* Make a copy of the key. The key is the entity name, and EntityLabelObject.EntityName
		requires a string reference, but we don't want to assign it the address of key, because
		key is a loop variable.
		*/
		entity_name := key
		value = strings.ToLower(value)
		start_index := int32(strings.Index(text, value))
		end_index := start_index + int32(len(value))
		if start_index > -1 {
			results = append(results, authoring.EntityLabelObject { EntityName: &entity_name, StartCharIndex: &start_index, EndCharIndex: &end_index })
		}
	}
	
	fmt.Printf("Created %d utterances.\n", len(results))

	return authoring.ExampleLabelObject { Text: &text, EntityLabels: &results, IntentName: &intent }
}

func add_utterances(app_id string, app_version string) {
	ctx := context.Background()

	client := authoring.NewExamplesClient(endpoint)
	// Set the subscription key on the client.
	client.Authorizer = autorest.NewCognitiveServicesAuthorizer(subscription_key)

	uuid_app_id, err := uuid.FromString(app_id)
    if err != nil {
		log.Fatal(err)
    }

	// Each entity must have at least one utterance before we train the application.
	utterance_1 := create_utterance("FindFlights", "find flights in economy to Madrid", map[string]string{ "Flight": "economy to Madrid", "Destination": "Madrid", "Class": "economy" })
	utterance_2 := create_utterance("FindFlights", "find flights to London in first class", map[string]string { "Flight": "London in first class", "Destination": "London", "Class": "first" })
	utterances := []authoring.ExampleLabelObject{ utterance_1, utterance_2 }

	_, err = client.Batch(ctx, uuid_app_id, app_version, utterances)
	    if err != nil {
		log.Fatal(err)
    }

	fmt.Println("Example utterances added.")
}

func train_app(app_id string, app_version string) {
	// Get the context, which is required by the SDK methods.
	ctx := context.Background()

	client := authoring.NewTrainClient(endpoint)
	// Set the subscription key on the client.
	client.Authorizer = autorest.NewCognitiveServicesAuthorizer(subscription_key)

	uuid_app_id, err := uuid.FromString(app_id)
    if err != nil {
		log.Fatal(err)
    }

	_, err = client.TrainVersion(ctx, uuid_app_id, app_version)
    if err != nil {
		log.Fatal(err)
    }

	// Wait for the train operation to finish.
	fmt.Println ("Waiting for train operation to finish...")

	waiting := true
	for true == waiting {
		result, err := client.GetStatus(ctx, uuid_app_id, app_version)
		if err != nil {
			log.Fatal(err)
		}
// GetStatus returns a list of training statuses, one for each model. Loop through them and make sure all are done.
		waiting = false
		for _, item := range *result.Value {
			details := *item.Details
			status := string (details.Status)
			if "Queued" == status || "InProgress" == status {
				waiting = true
			}
			if "Fail" == status {
				fmt.Println("Training operation failed. Result:")
				json_result, err := json.Marshal(result)
				if err != nil {
					log.Fatal(err)
				}
				log.Fatal(string(json_result))
			}
		}
		if true == waiting {
			fmt.Println("Waiting 10 seconds for training to complete...")
			time.Sleep(10 * time.Second)
		}
	}
}

func publish_app(app_id string, app_version string) {
	// Get the context, which is required by the SDK methods.
	ctx := context.Background()

	client := authoring.NewAppsClient(endpoint)
	// Set the subscription key on the client.
	client.Authorizer = autorest.NewCognitiveServicesAuthorizer(subscription_key)

	uuid_app_id, err := uuid.FromString(app_id)
    if err != nil {
		log.Fatal(err)
    }

	is_staging := true
	response, err := client.Publish(ctx, uuid_app_id, authoring.ApplicationPublishObject{ VersionID: &app_version, IsStaging: &is_staging })
    if err != nil {
		log.Fatal(err)
    }
	fmt.Println("Application published. Endpoint URL: " + *response.EndpointURL)
}

func delete_app(app_id string, app_version string) {
	// Get the context, which is required by the SDK methods.
	ctx := context.Background()

	client := authoring.NewAppsClient(endpoint)
	// Set the subscription key on the client.
	client.Authorizer = autorest.NewCognitiveServicesAuthorizer(subscription_key)

	uuid_app_id, err := uuid.FromString(app_id)
    if err != nil {
		log.Fatal(err)
    }

	force := false
	_, err = client.Delete(ctx, uuid_app_id, &force)
    if err != nil {
		log.Fatal(err)
    }
	fmt.Println("Application deleted.")
}

func list_apps() {
	// Get the context, which is required by the SDK methods.
	ctx := context.Background()

	client := authoring.NewAppsClient(endpoint)
	// Set the subscription key on the client.
	client.Authorizer = autorest.NewCognitiveServicesAuthorizer(subscription_key)

	var skip int32 = 0
	var max int32 = 100
	result, err := client.List(ctx, &skip, &max)
	if err != nil {
		log.Fatal(err)
	}
	for _, item := range *result.Value {
		fmt.Println(*item.ID)
	}

}

func main() {
	app_id, app_version := create_app()
	list_apps()
	add_entities(app_id, app_version)
	add_intents(app_id, app_version)
	add_utterances(app_id, app_version)
	train_app(app_id, app_version)
	publish_app(app_id, app_version)
	delete_app(app_id, app_version)
}

File: java/LUIS/authoring/quickstart.java
import com.microsoft.azure.cognitiveservices.language.luis.authoring.*;
import com.microsoft.azure.cognitiveservices.language.luis.authoring.models.*;

import java.io.*;
import java.lang.Object.*;
import java.time.format.DateTimeFormatter;  
import java.time.LocalDateTime; 
import java.util.*;
import java.net.*;

/* This sample for the Azure Cognitive Services LUIS API shows how to:
 * - Create an application.
 * - Add intents to an application.
 * - Add entities to an application.
 * - Add utterances to an application.
 * - Train an application.
 * - Publish an application.
 * - Delete an application.
 * - List all applications.
 */

/* To collect the Maven libraries needed, enter this at the command prompt:
 * mvn clean dependency:copy-dependencies
 * 
 * To compile and run, enter the following at a command prompt:
 * javac Quickstart.java -cp .;lib\*
 * java -cp .;lib\* Quickstart
 * 
 * This presumes your libraries are stored in a folder named "lib"
 * directly under the current folder. If not, please adjust the
 * -classpath/-cp value accordingly.
 */

/* Note If you run this sample with JRE 9+, you may encounter the following issue:
 * https://github.com/Azure/autorest-clientruntime-for-java/issues/569
 * which results in the following output:
 *
 * WARNING: An illegal reflective access operation has occurred
 * WARNING: Illegal reflective access by com.microsoft.rest.Validator (file:client-runtime-1.6.6.jar) to field java.util.ArrayList.serialVersionUID
 * WARNING: Please consider reporting this to the maintainers of com.microsoft.rest.Validator
 * WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
 * WARNING: All illegal access operations will be denied in a future release
 *
 * This should not prevent the sample from running correctly.
 */

public class Quickstart {
	/* Configure the local environment:
	* Set the following environment variables on your local machine using the
	* appropriate method for your preferred shell (Bash, PowerShell, Command
	* Prompt, etc.).
	*
	* LUIS_AUTHORING_KEY
	* LUIS_AUTHORING_ENDPOINT
	*
	* If the environment variable is created after the application is launched in a console or with Visual
	* Studio, the shell (or Visual Studio) needs to be closed and reloaded to take the environment variable into account.
	*/
    private static String authoring_key = System.getenv("LUIS_AUTHORING_KEY");
	private static String s_authoring_endpoint = System.getenv("LUIS_AUTHORING_ENDPOINT");

// Note EndpointAPI.fromString cannot parse a URL that begins with "https://", so we remove it.
	private static com.microsoft.azure.cognitiveservices.language.luis.authoring.EndpointAPI authoring_endpoint = com.microsoft.azure.cognitiveservices.language.luis.authoring.EndpointAPI.fromString(s_authoring_endpoint.replace ("https://", ""));

	private class AppInfo {
		public String app_id;
		public String app_version;
		public AppInfo(String app_id, String app_version) {
			this.app_id = app_id;
			this.app_version = app_version;
		}
	}

	public static void list_applications() {
		LUISAuthoringClient client = LUISAuthoringManager.authenticate(authoring_endpoint, authoring_key);

		Apps.AppsListDefinitionStages.WithExecute stages = client.apps().list();
		List<ApplicationInfoResponse> result = stages.execute();

		for (ApplicationInfoResponse app : result) { 
			System.out.print(app.id().toString()); 
		} 
	}

	public AppInfo create_application() {
		LUISAuthoringClient client = LUISAuthoringManager.authenticate(authoring_endpoint, authoring_key);

		DateTimeFormatter dtf = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");  
		String name = "Contoso " + dtf.format(LocalDateTime.now());
		String description = "Flight booking app built with Azure SDK for Java.";
		String version = "0.1";
		String locale = "en-us";
		ApplicationCreateObject create_app_payload = new ApplicationCreateObject().withName(name).withDescription(description).withInitialVersionId(version).withCulture(locale);
		UUID app_id = client.apps().add(create_app_payload);
		System.out.println("Created LUIS app " + name + " with ID " + app_id.toString());
		return new AppInfo(app_id.toString(), version);
	}

	public void add_entities(AppInfo app_info) {
		LUISAuthoringClient client = LUISAuthoringManager.authenticate(authoring_endpoint, authoring_key);

		String entity_name = "Destination";
		Models.ModelsAddEntityDefinitionStages.WithAppId stages = client.models().addEntity();
		stages.withAppId(UUID.fromString(app_info.app_id)).withVersionId(app_info.app_version).withName(entity_name).execute();

		String h_entity_name = "Class";
		List<String> h_entity_children = Arrays.asList("First", "Business", "Economy");
		client.models().addHierarchicalEntity(UUID.fromString(app_info.app_id), app_info.app_version, new HierarchicalEntityModel().withName(h_entity_name).withChildren(h_entity_children));

		String c_entity_name = "Flight";
		List<String> c_entity_children = Arrays.asList("Class", "Destination");
		client.models().addCompositeEntity(UUID.fromString(app_info.app_id), app_info.app_version, new CompositeEntityModel().withName(c_entity_name).withChildren(c_entity_children));

		System.out.println("Entities Destination, Class, Flight created.");
	}

	public void add_intents(AppInfo app_info) {
		LUISAuthoringClient client = LUISAuthoringManager.authenticate(authoring_endpoint, authoring_key);

		String intent_name = "FindFlights";
		Models.ModelsAddIntentDefinitionStages.WithAppId stages = client.models().addIntent();
		stages.withAppId(UUID.fromString(app_info.app_id)).withVersionId(app_info.app_version).withName(intent_name);

		System.out.println("Intent FindFlights added.");
	}

	public ExampleLabelObject create_utterance(String intent, String text, Map<String, String> labels) {
		text = text.toLowerCase();

		ArrayList<EntityLabelObject> entityLabels = new ArrayList<EntityLabelObject>();

		for (Map.Entry<String, String> entry : labels.entrySet()) {
			String key = entry.getKey();
			String value = entry.getValue().toLowerCase();
			
			int start_index = text.indexOf(value);
			int end_index = start_index + value.length();
			if (start_index > -1) {
				entityLabels.add(new EntityLabelObject().withEntityName(key).withStartCharIndex(start_index).withEndCharIndex(end_index));
			}
		}
	
		System.out.println("Created " + entityLabels.size() + " utterances.");

		return new ExampleLabelObject().withText(text).withEntityLabels(entityLabels).withIntentName(intent);
	}

	public void add_utterances(AppInfo app_info) {
		LUISAuthoringClient client = LUISAuthoringManager.authenticate(authoring_endpoint, authoring_key);

		// Each entity must have at least one utterance before we train the application.
		ExampleLabelObject utterance_1 = create_utterance("FindFlights", "find flights in economy to Madrid", Map.of("Flight", "economy to Madrid", "Destination", "Madrid", "Class", "economy"));
		ExampleLabelObject utterance_2 = create_utterance("FindFlights", "find flights to London in first class", Map.of ("Flight", "London in first class", "Destination", "London", "Class", "first"));
		List<ExampleLabelObject> utterances = Arrays.asList(utterance_1, utterance_2);

		client.examples().batch(UUID.fromString(app_info.app_id), app_info.app_version, utterances);

		System.out.println("Example utterances added.")	;
	}

	public void train_application(AppInfo app_info) throws InterruptedException {
		LUISAuthoringClient client = LUISAuthoringManager.authenticate(authoring_endpoint, authoring_key);

		client.trains().trainVersion(UUID.fromString(app_info.app_id), app_info.app_version);

		// Wait for the train operation to finish.
		System.out.println ("Waiting for train operation to finish...");

		Boolean waiting = true;
		while (true == waiting) {
			List<ModelTrainingInfo> result = client.trains().getStatus(UUID.fromString(app_info.app_id), app_info.app_version);
			// GetStatus returns a list of training statuses, one for each model.
			// Loop through them and make sure all are done.
			waiting = false;
			for (ModelTrainingInfo info : result) { 
				String status = info.details().status();
				if (status.equals("Queued") || status.equals("InProgress")) {
					waiting = true;
				}
				if (status.equals("Fail")) {
					System.out.println("Training operation failed. Reason: " + info.details().failureReason());
				}
			}
			if (true == waiting) {
				System.out.println("Waiting 10 seconds for training to complete...");
				Thread.sleep(10000);
			}
		}
	}

	public void publish_application(AppInfo app_info) {
		LUISAuthoringClient client = LUISAuthoringManager.authenticate(authoring_endpoint, authoring_key);

		 ProductionOrStagingEndpointInfo result = client.apps().publish(UUID.fromString(app_info.app_id), new ApplicationPublishObject().withVersionId(app_info.app_version).withIsStaging(true));
		
		System.out.println("Application published. Endpoint URL: " + result.endpointUrl());
	}

	public void delete_application(AppInfo app_info) {
		LUISAuthoringClient client = LUISAuthoringManager.authenticate(authoring_endpoint, authoring_key);

		OperationStatus status = client.apps().delete(UUID.fromString(app_info.app_id));
		OperationStatusType code = status.code();
		System.out.println("Deleted LUIS app with ID " + app_info.app_id + ". Operation status: " + code);
	}

    public static void main(String[] args) {
		try {
			Quickstart quickstart = new Quickstart();
			AppInfo app_info = quickstart.create_application();
			quickstart.add_entities(app_info);
			quickstart.add_intents(app_info);
			quickstart.add_utterances(app_info);
			quickstart.train_application(app_info);
			quickstart.publish_application(app_info);
			quickstart.delete_application(app_info);
        } catch (Exception e) {
            System.out.println(e.getMessage());
            e.printStackTrace();
        }
    }

}
